keps/sig-network/4770-endpointslice-controller-flexibility/README.md

```diff
@@ -0,0 +1,365 @@
+
+The custom EndpointSlice Controller will then read the annotation to select the pods targeted by the service.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: service
+  annotations:
+    selector: "app=a"
+spec: {}
+```
+
+This alternative potentially leads to confusion among users and inconsistency in how services are managed as each implementation is using its own annotation (see the nokia/danm and projectcalico/vpp-dataplane examples), leading to a fragmented approach.
+
+### Well-known label: Use Dummy Selector
+
+The set of Pods targeted by a Service is determined by a selector, the labels in the selector must be included as part of the pod labels. If a dummy selector is added to the service, Kubernetes will not select any pod, the endpointslices created by Kubernetes will then be empty. To simplify the user experience, a mutating webhook could add the dummy selector when the type of service is detected.
+
+The custom EndpointSlice Controller could read the service.spec.selector and ignore the dummy label to select pods targeted by the service.
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: service
+spec:
+  selector:
+    app: a
+    dummy-selector: "true"
+```
+
+This alternative fails to prevent the placeholder (empty) EndpointSlice(s) to be created by Kube-Controller-Manager. This also potentially causes confusion among users as every implementation could use a different dummy-selector key. Additionally, a miss-configuration with the missing dummy label will lead to unintended EndpointSlices being created with Pod.Status.PodIPs.
+
+### Well-known label: Disable the Kube-Controller-Manager Controllers
+
+The list of controllers to enable in the Kube-Controller-Manager can be set using the `--controllers` flag ([kube-controller-manager documentation](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/)). The EndpointSlice can then be disabled in the Kube-Controller-Manager and implemented as an external one that will support the label feature described in this KEP.
+
+This alternative requires significant changes to the cluster management as the cluster level configuration must be modified and a new EndpointSlice controller for the primary network must be developed and deployed to replace the disabled one in the Kube-Controller-Manager.
+
```

Comment:
We should - I pretty much think **must** - document the "don't specify a selector, then do your own thing" option. It _is_ an alternative and we shouldn't pretend otherwise.

Reply:
This is covered by the first alternative `Well-known label: Empty Selector field and Use Annotation as Selector`, I renamed the alternative so it is more clear.

Reply:
BTW you can't select on annotations.

Reply:
you can create your own selector in the annotation, here is an example.
```yaml
apiVersion: v1
kind: Service
metadata:
  name: service
  annotations:
    selector: "app=a"
spec: {}
```

Reply:
Noted